osutil/vfs/propagation.go

```diff
@@ -0,0 +1,274 @@
+//
+// The unbindable flag is always cleared.
+// If the mount point is not yet shared, it joins a new peer group as the only member.
+// When a shared mount is used as a bind-mount source, the shared peer group is retained.
+// When a shared mount dominates a mount or unmount operation, the operation is replayed to all the members of the
+// shared peer group as well as to all the slaves that were former members.
+func (v *VFS) MakeShared(mountPoint string) error {
+	return v.makeShared(mountPoint, 0)
+}
+
+// MakeRecursivelyShared is the recursive version of MakeShared.
+func (v *VFS) MakeRecursivelyShared(mountPoint string) error {
+	return v.makeShared(mountPoint, flagRecurse)
+}
+
+// MakeSlave makes a mount point a new member of a slave group.
+//
+// The unbindable flag is always cleared. If the mount is shared then it
+// converts from a peer to a slave of the group. Mounts that are slave to a
+// peer group receive mount and unmount operations from the group, but do not
+// propagate any operations back.
+func (v *VFS) MakeSlave(mountPoint string) error {
+	return v.makeSlave(mountPoint, 0)
+}
+
+// MakeRecursivelySlave is a recursive version of MakeSlave.
+func (v *VFS) MakeRecursivelySlave(mountPoint string) error {
+	return v.makeSlave(mountPoint, flagRecurse)
+}
+
+// MakePrivate removes a mount point from peer groups and makes it bindable again.
+func (v *VFS) MakePrivate(mountPoint string) error {
+	return v.makePrivate(mountPoint, 0)
+}
+
+// MakeRecursivelyPrivate is a recursive version of MakePrivate.
+func (v *VFS) MakeRecursivelyPrivate(mountPoint string) error {
+	return v.makePrivate(mountPoint, flagRecurse)
+}
+
+// MakeUnbindable makes the mount point unbindable.
+//
+// A bind mount operation that uses an unbindable mount as source immediately
+// fails. A recursive bind mount operation ignores any unbindable mounts without
+// failure.
+func (v *VFS) MakeUnbindable(mountPoint string) error {
+	return v.makeUnbindable(mountPoint, 0)
+}
+
+// MakeRecursivelyUnbindable is a recursive version of MakeUnbindable.
+func (v *VFS) MakeRecursivelyUnbindable(mountPoint string) error {
+	return v.makeUnbindable(mountPoint, flagRecurse)
+}
+
+// prpagationChange applies a propagation change function fn at a given mount point.
+//
+// The string [op] is used to construct an error value in case [mountPoint] is
+// not a mount point.  The [recurse] flag causes the function to be called
+// recursively on all the descendants of the mount denoted by [mountPoint]. The
+// descendants are all the mounts whose parentID is that can be traced back to
+// the selected starting mount.
+func (v *VFS) propagationChange(mountPoint, op string, recurse bool, fn func(m *mount)) error {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	return v.atMountPoint(mountPoint, op, func(pd pathDominator) error {
+		fn(pd.mount)
+
+		if recurse {
+			ancestorID := pd.mount.mountID
+			for _, m := range v.mounts[pd.index+1:] {
```

Comment:
This should use the new links instead of brute-force iteration. It should also not use index+1 as the starting point.